src/durmod/kaldi-durmod.h

```diff
@@ -0,0 +1,266 @@
+#include "util/stl-utils.h"
+#include "fstext/deterministic-fst.h"
+
+namespace kaldi {
+  using nnet3::Nnet;
+  using nnet3::NnetExample;
+  using nnet3::NnetIo;
+  using nnet3::CachingOptimizingCompiler;
+  using nnet3::ComputationRequest;
+  using nnet3::NnetComputation;
+  using nnet3::NnetComputeOptions;
+  using nnet3::NnetComputer;
+
+struct PhoneDurationModelOptions {
+  int32 left_context, right_context;
+  int32 max_duration;
+
+  void Register(OptionsItf *po);
+
+  PhoneDurationModelOptions():
+    left_context(4),
+    right_context(2),
+    max_duration(60) {}
+};
+
+class PhoneDurationModel {
+  friend class PhoneDurationFeatureMaker;
+
+ public:
+  void Read(std::istream &is, bool binary);
+  void Write(std::ostream &os, bool binary) const;
+
+  PhoneDurationModel() {}
+  PhoneDurationModel(const PhoneDurationModelOptions &opts,
+                     const std::vector<std::vector<int32> > &roots,
+                     const std::vector<std::vector<int32> > &questions):
+    left_context_(opts.left_context),
+    right_context_(opts.right_context),
+    roots_(roots),
+    questions_(questions),
+    max_duration_(opts.max_duration) {}
+
+  inline int32 MaxDuration() const { return max_duration_; }
+  inline int32 RightContext() const { return right_context_; }
+  inline int32 LeftContext() const { return left_context_; }
+  inline int32 FullContextSize() const {
+    return left_context_ + right_context_ + 1;
+  }
+  std::string Info() const;
+
+ private:
+  int32 left_context_, right_context_;
+  std::vector<std::vector<int32> > roots_;
+  std::vector<std::vector<int32> > questions_;
+  int32 max_duration_;
+};
+
+class PhoneDurationFeatureMaker {
+ public:
+  explicit PhoneDurationFeatureMaker(const PhoneDurationModel &model);
+
+  void InitFeatureMaker(const PhoneDurationModel &model);
+
+  /// This method extracts features for a phone. The inputs are a
+  /// phone-duration context and an index into that context which indicates the
+  /// middle phone (middle in the sense of left/right context).
+  /// a phone-duration context is a sequence of <phone-ID, duration-in-frames>
+  /// pairs.
+  /// The extracted features consist of phone-IDs (which are represented in a
+  /// 1-of-n encoding, suitable for neural networks), phone durations, and
+  /// binary features determined from a set of questions
+  /// (i.e. extra_questions.int)
+  void MakeFeatureVector(
+                 const std::vector<std::pair<int32, int32> > &phone_dur_context,
+                 int phone_index,
+                 SparseVector<BaseFloat> *feat) const;
+
+
+  int32 FeatureDim() const { return feature_dim_; }
+  int32 NumBinaryFeatures() const { return num_binary_features_; }
+  int32 NumPhoneIdentities() const { return num_phone_identities_; }
+  int32 OutputDim() const { return max_duration_; }
+
+  std::string Info() const;
+
+ private:
+  unordered_map<int32, std::vector<int32> > binary_feats_;
+  unordered_map<int32, int32> phone_id_;
```

Comment:
It's a glitch with the GridEngine queuing software.  We had some problems
today.
Just rerun that one manually if you can and continue.
Dan


On Fri, Feb 19, 2016 at 6:14 PM, Hossein Hadian <notifications@github.com>
wrote:

> In src/durmod/kaldi-durmod.h
> <https://github.com/kaldi-asr/kaldi/pull/414#discussion_r53532305>:
>
> > +  void MakeFeatureVector(
> > +                 const std::vector<std::pair<int32, int32> > &phone_dur_context,
> > +                 int phone_index,
> > +                 SparseVector<BaseFloat> *feat) const;
> > +
> > +
> > +  int32 FeatureDim() const { return feature_dim_; }
> > +  int32 NumBinaryFeatures() const { return num_binary_features_; }
> > +  int32 NumPhoneIdentities() const { return num_phone_identities_; }
> > +  int32 OutputDim() const { return max_duration_; }
> > +
> > +  std::string Info() const;
> > +
> > + private:
> > +  unordered_map<int32, std::vector<int32> > binary_feats_;
> > +  unordered_map<int32, int32> phone_id_;
>
> The error was this (of course the logfiles are not available now):
> queue.pl: Error, unfinished job no longer exists, log is in
> exp/lstm_bidirectional_ld0/durmod20_decode_eval_dm_rescored/log/rescore.17.log,
> last line is 'LOG
> (nnet3-durmodel-rescore-lattice:main():nnet3-durmodel-rescore-lattice.cc:77)
> Rescoring lattice for key AMI_EN2002a_MDM_MEE071_0119796_0120427', syncfile
> is
> exp/lstm_bidirectional_ld0/durmod20_decode_eval_dm_rescored/q/done.31693.17,
> return status of qstat was 256
> Possible reasons: a) Exceeded time limit? -> Use more jobs! b)
> Shutdown/Frozen machine? -> Run again!
>
> But this can not be from the memory, because I never had this problem on
> AMI before. The only difference now is that the neural network is bigger
> which does not affect the composed fst. Also I checked the memory usage of
> the jobs that took very long using htop and it was 153 MB with CPU
> utilization at 100%.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kaldi-asr/kaldi/pull/414/files#r53532305>.
>